source/logrepl/snapshot_test.go

```diff
@@ -0,0 +1,155 @@
+
+	type fields struct {
+		uri          string
+		table        string
+		columns      []string
+		keyColumn    string
+		snapshotName string
+	}
+	type args struct {
+		in0 context.Context
+	}
+	tests := []struct {
+		name    string
+		fields  fields
+		args    args
+		want    sdk.Record
+		wantErr bool
+	}{
+		{
+			name: "should return the first record",
+			fields: fields{
+				table:        test.SetupTestTable(ctx, t, db),
+				columns:      []string{"id", "key", "column1", "column2", "column3"},
+				keyColumn:    "key",
+				uri:          test.RepmgrConnString,
+				snapshotName: createTestSnapshot(t, db),
+			},
+			args: args{
+				in0: context.Background(),
+			},
+			want: sdk.Record{
+				Position: sdk.Position("0"),
+				Key: sdk.StructuredData{
+					"key": []uint8("1"),
+				},
+				Payload: sdk.StructuredData{
+					"id":      int64(1),
+					"column1": "foo",
+					"column2": int32(123),
+					"column3": bool(false),
+				},
+				Metadata: map[string]string{
+					"action": actionSnapshot,
+				},
+			},
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			s, err := NewSnapshotIterator(context.Background(), SnapshotConfig{
+				SnapshotName: tt.fields.snapshotName,
+				Table:        tt.fields.table,
+				Columns:      tt.fields.columns,
+				KeyColumn:    tt.fields.keyColumn,
+				URI:          tt.fields.uri,
+			})
+			i.NoErr(err)
+
+			t.Cleanup(func() {
+				i.NoErr(s.Teardown(context.Background()))
+			})
+
+			now := time.Now()
+			got, err := s.Next(tt.args.in0)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("SnapshotIterator.Next() error = %v, wantErr %v", err, tt.wantErr)
+			}
+
+			i.True(got.Metadata["table"] == tt.fields.table)
+			delete(got.Metadata, "table") // delete so we don't compare in diff
+			i.True(got.CreatedAt.After(now))
+			got.CreatedAt = time.Time{} // reset so we don't compare in diff
+
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("SnapshotIterator.Next() = %v, want %v", got, tt.want)
+				if diff := cmp.Diff(got, tt.want); diff != "" {
+					t.Log(diff)
+				}
+			}
+		})
+	}
+}
+
+func TestIteration(t *testing.T) {
+	i := is.New(t)
+	ctx := context.Background()
+	db := test.ConnectSimple(ctx, t, test.RepmgrConnString)
+
+	s, err := NewSnapshotIterator(context.Background(), SnapshotConfig{
+		SnapshotName: createTestSnapshot(t, db),
+		Table:        test.SetupTestTable(ctx, t, db),
+		Columns:      []string{"id", "key", "column1", "column2", "column3"},
+		KeyColumn:    "key",
+		URI:          test.RepmgrConnString,
+	})
+	i.NoErr(err)
+
+	now := time.Now()
+	for idx := 0; idx < 2; idx++ {
+		rec, err := s.Next(ctx)
+		i.Equal(string(rec.Position), strconv.FormatInt(int64(idx), 10))
+		i.NoErr(err)
+		t.Log(rec)
+		rec.CreatedAt.After(now)
+	}
+
+	i.NoErr(s.Teardown(ctx))
+}
+
+func createTestSnapshot(t *testing.T, db *pgx.Conn) string {
+	var n *string
+	query := `SELECT * FROM pg_catalog.pg_export_snapshot();`
```

Comment:
According to [Postgres docs](https://www.postgresql.org/docs/current/functions-admin.html#FUNCTIONS-SNAPSHOT-SYNCHRONIZATION):
> Saves the transaction's current snapshot and returns a text string identifying the snapshot. This string must be passed (outside the database) to clients that want to import the snapshot. **The snapshot is available for import only until the end of the transaction that exported it.**

So to set up the test correctly we need to 1. open a separate connection, 2. open a transaction in that connection and 3. export a snapshot. The transaction and connection should stay open at least until we create the snapshot iterator.